src/Compilers/Core/Portable/CommandLine/AnalyzerConfigSet.cs

```diff
@@ -562,52 +562,61 @@ private Section GetSection(string sectionName)
                 return new Section(sectionName, result);
             }
 
-            private void MergeSection(string configPath, Section section, bool isGlobalSection)
+            private void MergeSection(string configPath, Section section, int globalLevel, bool isGlobalSection)
             {
                 Debug.Assert(_values is object);
                 Debug.Assert(_duplicates is object);
 
                 if (!_values.TryGetValue(section.Name, out var sectionDict))
                 {
-                    sectionDict = ImmutableDictionary.CreateBuilder<string, (string, string)>(Section.PropertiesKeyComparer);
+                    sectionDict = ImmutableDictionary.CreateBuilder<string, (string, string, int)>(Section.PropertiesKeyComparer);
                     _values.Add(section.Name, sectionDict);
                 }
 
                 _duplicates.TryGetValue(section.Name, out var duplicateDict);
                 foreach ((var key, var value) in section.Properties)
                 {
-                    if (isGlobalSection && Section.PropertiesKeyComparer.Equals(key, GlobalKey))
+                    if (isGlobalSection && (Section.PropertiesKeyComparer.Equals(key, GlobalKey) || Section.PropertiesKeyComparer.Equals(key, GlobalLevelKey)))
                     {
                         continue;
                     }
 
                     bool keyInSection = sectionDict.ContainsKey(key);
                     bool keyDuplicated = duplicateDict?.ContainsKey(key) ?? false;
 
-                    // if this key is neither already present, or already duplicate, we can add it
-                    if (!keyInSection && !keyDuplicated)
+                    if ((!keyInSection && !keyDuplicated)   // this key is neither present nor a duplicate, we can add it
+                        || (keyInSection && sectionDict[key].globalLevel < globalLevel) // this key should override the existing section one
+                        || (keyDuplicated && duplicateDict![key].globalLevel < globalLevel)) // this key should override previous duplicates
                     {
-                        sectionDict.Add(key, (value, configPath));
+                        sectionDict[key] = (value, configPath, globalLevel);
```

Comment:
>sectionDict[key] = (value, configPath, globalLevel); [](start = 24, length = 52)

Is this correct if `sectionValue.globalLevel >= globalLevel` and `dictionaryValue.globalLevel < globalLevel`?

Reply:
We can't get into that situation: If we have one in the section, then it must be the highest precedence, and thus there isn't anything in the duplicate dictionary. If we have duplicates then they must be of the same precedence, and we would have removed the item from the section previously (which we assert at line 616)

Reply:
I don't quite follow. But does that mean if we find the key in the `sectionDict`, then we won't find it in `duplicateDict`? If so, would it help to make the assumption clear by only checking for duplicate if not found in the section?
```
bool keyInSection = sectionDict.TryGetValue(key, out var sectionValue);
bool keyDuplicated = !keyInSection && duplicateDict?.TryGetValue(key, out duplicateValue) == true;

if (!keyInSection || !keyDuplicated)
{
    sectionDict.Add(key, (value, configPath, globalLevel));
}
else
{
    int existingGlobalLevel = keyInSection ? sectionValue.globalLevel : duplicateValue.GlobalLevel;
    if (existingGlobalLevel < globalLevel)
    {
        ...
    }
    else if (existingGlobalLevel == globalLevel)
    {
        ...
    }
}
```

---
In reply to: [544743061](https://github.com/dotnet/roslyn/pull/49834#discussion_r544743061) [](ancestors = 544743061)

Reply:
Hmm, that is a _far_ nicer way of writing the same logic, thanks!